sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala

```diff
@@ -423,6 +345,121 @@ case class Murmur3Hash(children: Seq[Expression], seed: Int) extends Expression
       case udt: UserDefinedType[_] => computeHash(input, udt.sqlType, result, ctx)
     }
   }
+
+  protected def hasherClassName: String
+}
+
+/**
+ * Base class for interpreted hash functions.
+ */
+abstract class InterpretedHashFunction[E] {
```

Comment:
Should we just change this to always reutrn long and downcast where we currently use int?